### ISSUE TITLE
webp image rendering is much slower since the FreeImage.dll update in build 2023.9.30.299

### ISSUE THREAD

**Issue by Zyzyl** [Yellow-Dog-Man/Resonite-Issues#874]
### Describe the bug?

Rendering webp format images is surprisingly slow when using the up-to-date patched FreeImage.dll library packaged with the public build of Resonite. Using another (older and insecure) version packaged with the Steam build of Neos results in significantly faster rendering times. 

As a result, taking images using the default lossless image format feels less responsive than many users expect based on other similar platforms.

### To Reproduce

This issue may be investigated in this simple world: resrec:///U-Zyzyl/R-511f7ece-5f77-497e-ae52-335df587dc93
The world features ProtoFlux setups which can measure the time needed to render image via the RenderToTextureAsset node or the InteractiveCamera component.

I tested the 'new' version of FreeImage.dll using the current build of Resonite. I tested the 'old' version by replacing all of the FreeImage.dll instances in the Resonite install folder with the version available from the Steam build of Neos (2022.1.28.1335).

Using the 'new' version, I recorded the following average timings to render a 1920x1080 webp image (mean of 10 replicates):
- RenderToTextureAsset = 7.59 s
- InteractiveCamera = 7.55 s

Using the 'old' version:
- RenderToTextureAsset = 1.32 s
- InteractiveCamera = 1.27 s

### Expected behavior

Ideally webp image rendering would be as fast as it was before the security patch implemented in build 2023.9.30.299. 

Of course, this may not be possible if this degree of slowdown is critically important for secure handling of webp images.

### Resonite Version Number

2023.11.28.1365

### What Platforms does this occur on?

Windows

### What headset if any do you use?

Tested in Desktop mode

### Log Files

Log where I loaded the reproduction world and triggered the RenderToTextureAsset flux followed by the InteractiveCamera capture flux. This log using the 'new' FreeImage.dll as packed with the current build of Resonite.

[DESKTOP-SIS8AVA - 2023.11.28.1365 - 2023-12-11 22_19_29.log](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/13641043/DESKTOP-SIS8AVA.-.2023.11.28.1365.-.2023-12-11.22_19_29.log)

### Additional Context

Previously there was confusion about the source of the percieved slow image rendering in https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/871

### Reporters

zyzyl (zyzylian on Discord)

Involved with related issue https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/871
Ohzee
Virgil
Shifty

**Comment by Frooxius** [comment]
Thanks for the info!

I'll try to look into this when I can. But if anyone is interested in poking around, our fork or FreeImage is available here, which should include the changes made for the security fix in the Git history: https://github.com/Yellow-Dog-Man/FreeImage

**Comment by ohzee00** [comment]
https://github.com/Yellow-Dog-Man/FreeImage/commit/f53a9da7c9448b4b62e3cf38eaabd06b4563518b#diff-8ed1ba88d58aee947811598f8934862e15c12988c47791467de7a60a0ed161f0

This could very well be a "I have no idea what I'm talking about" moment but, could this change messed with screenshot encoding speed?

**Comment by Frooxius** [comment]
It could, but not sure why would that affect it. With new libWebP 6 is not valid value, it's either 0 or 1. 6 was also supposed to be slower.

**Comment by TisFoolish** [comment]
if `config.method` is compression strength and it goes up to six, what about testing other numbers? EG 4 or 5? I've noticed with PNG that you can get pretty high compression but still have it compress and save really fast if you don't set the compression level to its max value.

**Comment by Nutcake** [comment]
I did try both 4 and 5 as well, both resulting in about 5 seconds, I did not check file size with those though.

**Comment by Frooxius** [comment]
Those are good findings! I think the images are small enough anyways overall, that the quicker speed is more important here than the potential size savings - especially since the actual textures that will be loaded are re-compressed anyways.

One thing I find a bit odd here is that the slow down happened with the method not changing between the versions.

Would you be willing open a PR on the FreeImage with this? We can merge it in and push out in the next build.

**Comment by Nutcake** [comment]
Done.

**Comment by Frooxius** [comment]
Merged this in 2024.1.24.834! Thank you!

**Comment by goodusername123** [comment]
> One thing I find a bit odd here is that the slow down happened with the method not changing between the versions.

This happened because of this this commit https://github.com/Yellow-Dog-Man/FreeImage/commit/e155c2c6653f21cbf3494777207fa5abb82a63e1 which ended up activating the brute force lossless cruncher in libwebp.
The `quality` value has a very different meaning when lossless encoding is activated and does **not** relate to anything visual as both `method` and `quality` are used in conjunction to tune the effort/speed for lossless encoding.

**Comment by jonnyawsom3** [comment]
It's an old issue, but me and a friend recently did more testing around this in light of some strange behaviour with my screenshot saving and I feel like this is a good a place as any to give an update. There also seems to be a severe issue with PNG saving, but I'll open a new issue to track that.

TL;DR [this](https://github.com/Yellow-Dog-Man/FreeImage/pull/17) PR would further increase speed by around 25%, while also being 10% smaller. The speed increases with resolution, so for an 8K file it's 50% faster. The PR also changes the settings for lossy WebP separately, increasing session preview and thumbnail quality with a 15%~ filesize reduction.

Current preview quality, note the severe blocking and ringing around edges. 81.5 KB
![Current](https://github.com/user-attachments/assets/7d196121-27ce-4d5d-a42f-e3fcb42f7c81)
Our improved settings, giving a much cleaner image and more accurate colours. 67.9 KB
![Improved](https://github.com/user-attachments/assets/02ba823a-bd8e-4381-ad2d-7d51cfd6be2d)
It's a direct upgrade in all cases, so I wanted to bring it up again. Apologies for the necro.

**Comment by ProbablePrime** [comment]
I am re-opening this issue following a DM conversation with Jonny and due to the dependency(FreeImage) library PR referenced.

Im very busy so this is going straight to the backlog, but I will eventually make my way here and use all my new found image wizardness to objectively evaluate this in a similar way to my Compressonator fixes.

**Comment by jonnyawsom3** [comment]
~~Using the test world in the original comment:~~
> ~~This issue may be investigated in this simple world: resrec:///U-Zyzyl/R-511f7ece-5f77-497e-ae52-335df587dc93 The world features ProtoFlux setups which can measure the time needed to render image via the RenderToTextureAsset node or the InteractiveCamera component.~~

Turns out the mod only works for quick photos. Tests are done by comparing system time of photo taken vs time written.

Here are current PNG saving times compared to my recommendations in this issue https://github.com/Yellow-Dog-Man/FreeImage/issues/20, using a mod created by @yoshiyoshyosh.

Seconds | Old | New | Size+
| :---: | :---: | :---: | :---: |
1080p | 7 | <1 | 10%
1440p | 14 | <1 | 10%
2160p | 30 | 2 | 14%
4320p | 115 | 6 | 20%

Unfortunately testing WebP requires recompiling FreeImage, which multiple friends failed due to compile errors, but the improvements should be similar while still being smaller than Old PNG.
